### ISSUE TITLE
Request: User Controller Device

### ISSUE THREAD

**Issue by ProbablePrime** [Neos-Metaverse/NeosPublic#2554]
## Is your feature request related to a problem? Please describe.
Within Logix, we have User Head Output device, which can be useful to determine behavior based on a particular headset. 

We'd like to do this with controllers too, but currently the only way appears to be a slightly weird hack involving a Type Output on Standard Controller.

This requires a lot of nodes to get a full list etc. You can see a sample below. I apologize for the lack of resoloution:
![image](https://user-images.githubusercontent.com/8791132/123834275-1607e700-d8bc-11eb-92e7-7b06e1c47084.png)

## Relevant issues
I couldn't find any

## Describe the solution you'd like
A new node User Controller Device or similar which reports the type of controller a user is using. To emulate Head Output device this could be an Enum but I understand that there are a lot of types at play here that already exists.

## Describe alternatives you've considered
There's currently a tool within Essential -> Tools -> Misc that does this but looking into it. It uses weird substring matching which i feel is brittle and may break.

## Additional context
Requested by Epsilion, I'm Erin and myself.

**Comment by 3x1t-5tyl3** [comment]
"AvatarControllerInfo" somewhat does something similar. Though it outputs a "Type".
Thought I'd mention that.

**Comment by Aerizeon** [comment]
@3x1t-5tyl3 That's the same thing given by the Standard Controller Node.

The problem is that:
1. You have to maintain your own list of names mapped to types for a given controller, which isn't easily manageable (as shown above) 
2. It's not possible to differentiate between similar controller types (Rift S and CV1 touch controllers both output the same type since they have the same mapping, but have a vastly different physical appearance) - The same may apply for other controllers in the future.

It would be nice to have a node (or component) that provides the specific controller type, and an accurate presentable name.

**Comment by 3x1t-5tyl3** [comment]
Yeah I didn't think that far. Apologies. I just reminded myself of that component so I thought I should probably share that as that is somewhat related to what was being talked about :)

**Comment by Frooxius** [comment]
Type is the correct output here, the controller type cannot be an enum, because not all controller types might necessarily be known in advance at compile time (and thus put into the enum), they're added and initialized dynamically. It also ensures that if the type changes or is subclassed, compatibility gets maintained properly across versions. I'd also recommend using the Type input as well, rather than matching the substrings for this reason.

The HeadOutputDevice doesn't actually give you the type of the headset, it gives you the type of the runtime. You'll get the same value for lots of different headsets (e.g. SteamVR) and for some headsets you can get different ones depending on runtime (e.g. with Rift, you can run through both native Oculus Runtime and SteamVR).

I don't recommend using this to determine the actual type of the headset or do much custom behavior in the first place, as it might change in the future as runtimes change (e.g. say we switch to OpenXR for everything). There are other higher level nodes and mechanisms to check specific things (e.g. if user is currently in screen mode).

@Aerizeon The "ControllerDeviceModel" property on that component will actually let you differentiate between CV1 and RiftS controller types.

**Comment by ProbablePrime** [comment]
Thanks, @Frooxius based on that it sounds like the left hand side of my picture above is how you'd do this? We have most controller types there and are trying to find them.

Would a "list of types" input be possible? It would just have a list of types that can be scrolled through in a similar way to an Enum input but without actually being one under the hood?

The nodes on the left in this original screenshot are a lot of work and I don't expect anyone to do it correctly as there are so many

**Comment by Frooxius** [comment]
@ProbablePrime It depends on why are you doing this in the first place. Why are you trying to determine which controller is the user using?

But generally yes, if you want to handle each possible controller yourself, you'll have the burden of having to add specific checks for each type. It's potentially brittle approach though, because as we get more controllers and VR devices, you'll have to add handling for more and more of them yourself.

Instead I'd generally recommend using the abstracted mechanisms and interfaces where possible.

**Comment by ProbablePrime** [comment]
By abstracted you mean things like standard controller etc? That's a great node for a lot of usecase.

The exact scenario here was to display the 3D model of the controller from Neos Essentials ->  Controller layouts locally for each user.

This is similar to the component/facet that does it in the dash but in 3D in a world as a model. We eventually got this hooked up to a boolean switcher. 

Its a bit of a specific scenario where adding new controllers means new models and therefore new Type Inputs but we wanted to open this issue to discuss how we could potentially make this kind of scenario easier.

**Comment by Frooxius** [comment]
Yes, Standard Controller is one of the examples. Once we have exposed the Input Binding system it'll be another good system to use for handling inputs in a more generalized and flexible way. There are other behaviors as well for particular interactions, most of them should be controller-agnostic.

I see with the 3D model. Yeah in that case if you want to load a custom object for each, the responsibility of managing all controller types and matching objects to them falls to you. We could add component that will always load the default device model (NeosAvatarControllerSpawner does this specifically for avatar), but you wouldn't be able to override it with custom things.


For filtering inputs, that'll be possible with a proper editor/search for input types. E.g. we could mark given type to be anything that inherits from particular class and search for all currently loaded classes that match that description. Perhaps having helper methods to get common information based on type too once we have more support for that - e.g. getting canonical name, enumerating inputs and such.
